lib/rust/profiler-macros/src/lib.rs

```diff
@@ -0,0 +1,303 @@
+                let start = None;
+                let data = ProfilerData { parent, start, label };
+                Started { profiler, data }
+            }
+        }
+    };
+    ts.into()
+}
+
+/// impl Parent<T> for U, when U is an profiler of a disabled level.
+#[allow(non_snake_case)]
+fn disabled_impl_parent(
+    ParentIdent: &syn::Ident,
+    ChildIdent: &syn::Ident,
+) -> proc_macro::TokenStream {
+    let ts = quote::quote! {
+        impl Parent<#ChildIdent> for #ParentIdent {
+            fn new_child(&self, label: Label) -> Started<#ChildIdent> {
+                self.new_child_same_start(label)
+            }
+            fn new_child_same_start(&self, _label: Label) -> Started<#ChildIdent> {
+                let profiler = #ChildIdent(());
+                let data = ();
+                Started { profiler, data }
+            }
+        }
+    };
+    ts.into()
+}
+
+/// Return the numeric Profiling Level (counting from 0 = top-level only).
+fn get_enabled_level(levels: &[impl AsRef<str>]) -> usize {
+    const PROFILING_LEVEL_ENV_VAR: &str = "ENSO_MAX_PROFILING_LEVEL";
+    let enabled = match env::var(PROFILING_LEVEL_ENV_VAR) {
+        Ok(level) => level,
+        // If the variable isn't set, we default to the minimum.
+        Err(_) => return 0,
+    };
+    for (i, name) in levels.iter().enumerate() {
+        if &enabled[..] == name.as_ref() {
+            return i;
+        }
+    }
+    panic!("{} set to unknown profiling level: {}", PROFILING_LEVEL_ENV_VAR, enabled);
+}
+
+/// Defines a hierarchy of profiler levels.
+#[proc_macro]
+#[allow(non_snake_case)]
+pub fn define_hierarchy(ts: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let parser = punctuated::Punctuated::<syn::Ident, syn::Token![,]>::parse_terminated;
+    let idents: Vec<_> = parser.parse(ts).unwrap().into_iter().collect();
+    let level_names: Vec<_> = idents.iter().map(|id| id.to_string().to_snake_case()).collect();
+    let enabled_level = get_enabled_level(&level_names);
+    let mut out = proc_macro::TokenStream::new();
+    for ((i, ObjIdent), level_name) in idents.iter().enumerate().zip(level_names.iter()) {
+        let snake_ident = syn::Ident::new(level_name, proc_macro2::Span::call_site());
+        let profiler_def = match enabled_level >= i {
+            true => enabled_profiler(&snake_ident, ObjIdent),
+            false => disabled_profiler(&snake_ident, ObjIdent),
+        };
+        out.extend(profiler_def);
+        for (j, ChildIdent) in idents[i..].iter().enumerate() {
+            let parent_impl = match enabled_level >= i + j {
+                true => enabled_impl_parent(ObjIdent, ChildIdent),
+                false => disabled_impl_parent(ObjIdent, ChildIdent),
+            };
+            out.extend(parent_impl);
+        }
+    }
+    out
+}
+
+
+
+// ================
+// === profile! ===
+// ================
+
+/// Instruments a function.
+#[proc_macro_attribute]
+pub fn profile(
```

Comment:
Such macros should have a much more detailed documentation, especially in the situation when IntelliJ / CLion are unable to expand them in-place. This should have a description of what it does, what are its usages, what are its limitations (like no support for arg pattern matches), and also, it should show the expanded function for a particular  example function.